### PR TITLE
feat(optimization): Try to optimize ownership of assign statements

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/8846

## Summary\*

With this we go from 300 array copies down to 3 (not scaling with loop iterations) for the given test program.
Note that the example program uses slices so https://github.com/noir-lang/noir/pull/8867 is necessary to count the slice copies.

## Additional Context

This is a draft while I verify some corner cases with the assign check.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by jfecher** [comment]
This PR should be ready for review now, I can't seem to find any bugs with the approach here.

File: compiler/noirc_frontend/src/ownership/mod.rs [hunk 1]
```diff
@@ -320,6 +320,19 @@ impl Context {
     }
 
     fn handle_call(&mut self, call: &mut crate::monomorphization::ast::Call) {
+        // Hack: We don't have references yet but pretend the array length builtin
```
Comment (vezenovm) [inline compiler/noirc_frontend/src/ownership/mod.rs]:
```suggestion
        // Hack: We don't have references of `self` yet but pretend the array length builtin
```
A bit pedantic but would this not be more accurate?

Also do we have an issue for references to self we want to link here?

**Review by vezenovm — APPROVED** [review]

File: compiler/noirc_frontend/src/ownership/mod.rs (hunk 1 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/ownership/mod.rs]: That'd be less accurate - I mean we don't have immutable references in general currently. We do have `&mut self` which could be used but we don't want array len to take a mutable self reference. So it takes a `self` but we don't actually want to inc_rc just for a .len() call, hence this hack.

Comment (vezenovm) [reply compiler/noirc_frontend/src/ownership/mod.rs]: > we don't have immutable references in general currently.

Ah yes. I was confused with the wording as I was thinking we do have _mutable_ references.

**Comment by jfecher** [comment]
Found a couple of issues with this after thinking of some more cases where the previous value should not be moved. Just checking the loop index is not sufficient since:

```noir
for _ in 0 .. 10 {
    foo(bar);
}
for _ in 0 .. 10 {
    bar = &[];
}
```
In the above example if we just check the loop index, it will be `1` in both cases but we don't want to make `foo(bar)` a move since on the next iteration we'd be using a moved value.

**Comment by jfecher** [comment]
Abandoning this work for now to focus more on fixing bugs before 1.0.
This requires more than a day's work to fix it seems since I think we'll need some new tracking structures to track when this optimization is valid since the existing loop_index and branches structures are insufficient.
